### PR TITLE
test: switch to vitest

### LINKED ISSUE #5986: [Test] Vitest
Branch: [feat/vitest](https://github.com/element-plus/element-plus/tree/feat/vitest)

Gradually switch to Vitest

- Migrate utils and hooks
- Migrate component (in individual PR)

### PR DESCRIPTION
**Pull request by sxzz**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

I migrated `hooks` and `utils` to Vitest, and the rest of `components` part will plan to migrate in each separate PR.

### CODE REVIEW

**Comment by sxzz** [comment]
Here is some evidence that Vitest is reliable. /cc @JeremyWuuuuu 

- The one of maintainer of Vitest `patak_dev` is working **full-time** on Vitest. 
   - https://twitter.com/patak_dev/status/1491791394511351813
- Vitepress is using Vitest.
   - https://twitter.com/youyuxi/status/1483272831945248768
- There is now a @vitest_dev recipe in the new @vuejs docs! And it is recommended in the official scaffolding tool `pnpm create vue`.
   - https://vuejs.org/guide/scaling-up/testing.html#adding-vitest-to-a-project
   - https://twitter.com/vitest_dev/status/1490612354211528704
- There are some packages using Vitest
   - [faker.js](https://github.com/faker-js/faker/commit/60c90028ba76b7e291fdb8152425b93c41b117c9)
   - [nuxt 3](https://github.com/nuxt/framework/pull/2694)

File: packages/hooks/__tests__/use-locale.vitest.tsx [hunk 1]
```diff
@@ -1,46 +1,43 @@
...
-          locale: Object,
+          locale: Object as PropType<Language>,
         },
         components: {
           'el-test': TestComp,
         },
         setup(props, { slots }) {
           provideGlobalConfig(computed(() => ({ locale: props.locale })))
-          return () => slots.default()
+          return () => slots.default?.()
```
Comment (JeremyWuuuuu) [inline packages/hooks/__tests__/use-locale.vitest.tsx:35]: **Question:**  Can we use `jsx` in here, just to keep consistency.

**Review by jw-foss — COMMENTED** [review]
Most of the code made sense to me, I do have one question though, how should we collect the coverage and get the report?

File: packages/hooks/__tests__/use-locale.vitest.tsx (hunk 1 above)
Comment (sxzz) [reply packages/hooks/__tests__/use-locale.vitest.tsx:35]: I don't think it's necessary, it's just one line. If we use JSX, it just adds a bunch of angle brackets to the existing code.
```tsx
<>slots.default?.()</>
```

File: tsconfig.json [hunk 2]
```diff
@@ -11,13 +11,11 @@
     "allowSyntheticDefaultImports": true,
     "types": ["unplugin-vue-define-options"]
   },
-  "references": [{ "path": "./tsconfig.jest.json" }],
+  "references": [
+    { "path": "./tsconfig.vite-config.json" },
+    { "path": "./tsconfig.jest.json" },
+    { "path": "./tsconfig.vitest.json" }
+  ],
```
Comment (JeremyWuuuuu) [inline tsconfig.json:18]: **Praise:**  This was nice!

File: packages/hooks/__tests__/use-timeout.vitest.ts [hunk 3]
```diff
@@ -20,8 +20,9 @@ describe('use-timeout', () => {
     wrapper = _mount(cb)
   })
   afterEach(() => {
-    vi.useRealTimers()
     vi.clearAllMocks()
+    vi.restoreAllMocks()
```
Comment (JeremyWuuuuu) [inline packages/hooks/__tests__/use-timeout.vitest.ts:24]: I think `clearAllMocks` and `restoreAllMocks` are essentially functionally duplicated, can we only keep one of them?

**Review by jw-foss — COMMENTED** [review]
I left a comment for you to consider ;)

File: packages/hooks/__tests__/use-timeout.vitest.ts (hunk 3 above)
Comment (sxzz) [reply packages/hooks/__tests__/use-timeout.vitest.ts]: Fixed

**Review by jw-foss — APPROVED** [review]

**Pull request merged**